aten/src/ATen/Parallel.cpp

```diff
@@ -0,0 +1,63 @@
+#include <ATen/Parallel.h>
+
+#include <atomic>
+
+#ifdef TH_BLAS_MKL
+#include <mkl.h>
+#endif
+
+namespace at {
+
+namespace {
+// Number of threads set by the user
+std::atomic<int> num_threads(-1);
+}
+
+void init_num_threads() {
+  auto nthreads = num_threads.load();
+  if (nthreads > 0) {
+    set_num_threads(nthreads);
+  } else {
+#if defined(_OPENMP) && defined(TH_BLAS_MKL)
+  // If we are using MKL an OpenMP make sure the number of threads match.
+  // Otherwise, MKL and our OpenMP-enabled functions will keep changing the
+  // size of the OpenMP thread pool, resulting in worse performance (and memory
+  // leaks in GCC 5.4)
+  omp_set_num_threads(mkl_get_max_threads());
+#endif
+  }
+}
+
+void set_num_threads(size_t nthreads) {
+  if (nthreads == 0) {
+    return;
+  }
+#ifdef _OPENMP
+  omp_set_num_threads(nthreads);
+  num_threads.store(nthreads);
```

Comment:
shouldn't it be outside of #ifdef ?

Reply:
we could put this outside, though it will never be used if there's no parallelization library available (get_num_threads will always return 1); it all becomes no-ops essentially